nmostesting/mocks/Auth.py

```diff
@@ -0,0 +1,535 @@
+# Copyright (C) 2019 Advanced Media Workflow Association
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import json
+import uuid
+import flask
+import urllib
+import socket
+import base64
+import ssl
+import threading
+import os
+import jsonschema
+import requests
+
+from flask import Flask, Blueprint, Response, request, jsonify, redirect
+from urllib.parse import parse_qs
+from ..Config import PORT_BASE, KEYS_MOCKS, ENABLE_HTTPS, CERT_TRUST_ROOT_CA, JWKS_URI, REDIRECT_URI, SCOPE
+from ..TestHelper import get_default_ip, get_mocks_hostname, load_resolved_schema
+from ..IS10Utils import IS10Utils
+from zeroconf_monkey import ServiceInfo
+from enum import Enum
+from werkzeug.serving import make_server
+from http import HTTPStatus
+from authlib.jose import jwt
+
+
+GRANT_TYPES = Enum("grant_types", [
+    "authorization_code", "implicit", "refresh_token", "password", "client_credentials"])
+SCOPES = Enum("scopes", ["connection", "node", "query", "registration", "events", "channelmapping"])
+CODE_CHALLENGE_METHODS = Enum("code_challenge_methods", ["plain", "S256"])
+TOKEN_ENDPOINT_AUTH_METHODS = Enum("token_endpoint_auth_methods", [
+    "none", "client_secret_post", "client_secret_basic", "client_secret_jwt", "private_key_jwt"])
+RESPONSE_TYPES = Enum("response_types", ["code", "token"])
+
+
+class AuthException(Exception):
+    def __init__(self, error, description, httpstatus=HTTPStatus.BAD_REQUEST):
+        self.httpstatus = httpstatus
+        self.error = error
+        self.description = description
+
+
+class AuthServer(object):
+    def __init__(self, auth):
+        self.server = None
+        self.server_thread = None
+        self.auth = auth
+
+    def start(self):
+        """Start Authorization server"""
+        if not self.server_thread:
+            ctx = None
+            # placeholder for the certificate
+            cert_file = "test_data/BCP00301/ca/mock_auth_cert.pem"
+            # placeholder for the private key
+            key_file = "test_data/BCP00301/ca/mock_auth_private_key.pem"
+            # generate RSA key and certificate for the mock secondary Authorization server
+            IS10Utils.make_key_cert_files(cert_file, key_file)
```

Comment:
When the secondary auth server is started, does this overwrite files created for the primary auth server?? Does that matter?

Reply:
The primary Auth server is using the `CERTS_MOCKS` and `KEYS_MOCKS`, not the`test_data/BCP00301/ca/mock_auth_cert.pem` and `test_data/BCP00301/ca/mock_auth_private_key.pem`

Reply:
OK, that isn't clear to me how that happens... the primary one is `start`ed with `self.server_thread` but the secondary one isn't? :-/

Reply:
The primary auth server is started together with all other mocks inside the `start_web_servers` in the NMOSTesting.py, which all use the `CERTS_MOCKS` and the `KEYS_MOCKS`.

The secondary auth server use of the AuthServer to start and shut down the server. This use of the `test_data/BCP00301/ca/mock_auth_cert.pem` and `test_data/BCP00301/ca/mock_auth_private_key.pem` to start the server.

Reply:
OK, I understand, thanks.